### PR TITLE
Fix for NSD Attachment blocking

### PR DESCRIPTION
**Pull request by ctsims**
Workaround for NSD Services being a disaster and resulting in many-second blocking calls. Now will abandon the search after a brief window (it won't succeed after it's gotten into the bugged state anyway)

Ticket:
http://manage.dimagi.com/default.asp?222766#1120157

### CODE REVIEW

File: app/src/org/commcare/android/nsd/NSDDiscoveryTools.java [hunk 1]
```diff
@@ -106,6 +113,41 @@ private static void doDiscovery(Context context) {
...
+
+            boolean connected = new TimeBoundOperation(500) {
+                NsdManager manager;
+
+                @Override
+                public void run() {
+                    manager = (NsdManager)context.getSystemService(Context.NSD_SERVICE);
+                }
+
+                public void commit() {
```
Comment (phillipm) [inline app/src/org/commcare/android/nsd/NSDDiscoveryTools.java]: Needs an `@Override`

File: app/src/org/commcare/utils/TimeBoundOperation.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            }
+        });
+
+        thread.start();
+        try {
+            thread.join(timeout);
+        }catch (InterruptedException e) {
+            //Execution failed
+            return false;
+        };
```
Comment (phillipm) [inline app/src/org/commcare/utils/TimeBoundOperation.java]: unneeded `;`

File: app/src/org/commcare/utils/TimeBoundOperation.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+ *
+ * A time bound operation can only be executed once.
+ *
+ * Created by ctsims on 4/4/2016.
+ */
+public abstract class TimeBoundOperation {
+
+    final private long timeout;
+
+    private Thread thread;
```
Comment (phillipm) [inline app/src/org/commcare/utils/TimeBoundOperation.java]: can be local to `execute`

File: app/src/org/commcare/utils/TimeBoundOperation.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
+package org.commcare.utils;
+
+/**
+ * A time bound operation will run a synchronized operation in the background, with the confidence
```
Comment (phillipm) [inline app/src/org/commcare/utils/TimeBoundOperation.java]: "in the background": isn't it blocking the current thread? If so, maybe say something like "on a new, blocking thread"

File: app/src/org/commcare/android/nsd/NSDDiscoveryTools.java [hunk 5]
```diff
@@ -37,7 +41,8 @@
     public enum NsdState {
         Init,
         Discovery,
-        Idle
+        Idle,
+        Unsupported
```
Comment (phillipm) [inline app/src/org/commcare/android/nsd/NSDDiscoveryTools.java]: Might want to check for `Unsupported` in `addListener`

Comment (ctsims) [inline app/src/org/commcare/android/nsd/NSDDiscoveryTools.java]: We actually won't know when we add the listener, which seems fine. The caller is asking to subscribe to events that will never happen, but I don't think that should cause any lifecycle issues.

**Comment by ctsims** [comment]
@phillipm Feedback addressed, thanks for review!

**Pull request merged**